### PR TITLE
ido look

### PR DESCRIPTION
**Pull request by tomervak**

### CODE REVIEW

File: job-application-tracker-api/src/users/dto/login-user.dto.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import { IsEmail, IsNotEmpty } from "class-validator";
+
+export class loginDto{
```
Comment (ido-rosenberger) [inline job-application-tracker-api/src/users/dto/login-user.dto.ts]: Change to LoginRequest/LoginDto (class names always start with upper case letter)

use ctrl+r+r in your mother

File: job-application-tracker-api/src/users/users.controller.ts [hunk 2]
```diff
@@ -25,6 +27,12 @@ export class UsersController {
         return this.usersService.create(createUserDto);
     }
 
+    @Post('login')
+    async login(@Body() loginDto:loginDto) : Promise<User>{
```
Comment (ido-rosenberger) [inline job-application-tracker-api/src/users/users.controller.ts]: APIs should never return domain models. return a dto (Fix service login method comments first)

File: job-application-tracker-api/src/users/users.service.ts [hunk 3]
```diff
@@ -1,65 +1,85 @@
...
 import { UpdateUserDto } from './dto/update-user.dto';
 import { NotFoundException } from '@nestjs/common';
+import { loginDto } from './dto/login-user.dto';
+import { compare, hash } from 'bcrypt';
+import { User } from './dto/user.dto';
 
 @Injectable()
 export class UsersService {
-    private users = [
+    private users : User[] = [
```
Comment (ido-rosenberger) [inline job-application-tracker-api/src/users/users.service.ts]: Passwords in this array should be hashed (thought temporary)

File: job-application-tracker-api/src/users/users.service.ts [hunk 4]
```diff
@@ -1,65 +1,85 @@
...
             "firstName": "don",
             "lastName": "fil",
             "email": "bonfil98@gmail.com",
-            "password": "123456"
+            "passwordHash": "123456"
         }
-    ]
+    ];
 
+    userResponse(user :User) {
```
Comment (ido-rosenberger) [inline job-application-tracker-api/src/users/users.service.ts]: Change to "getUserResponse"
use ctrl+r+r in your mother

also add above the function decleration a comment //TODO: replace with a mapper

File: job-application-tracker-api/src/users/users.service.ts [hunk 5]
```diff
@@ -68,4 +88,17 @@ export class UsersService {
         this.users = this.users.filter(user => user.id !== id);
         return removedUser;
     }
+
+    async login(loginDto : loginDto): Promise<User>{
+        const user = await this.users.find(user => user.email===loginDto.email);
+        if(!user){
+            throw new HttpException('User not found',HttpStatus.UNPROCESSABLE_ENTITY);
```
Comment (ido-rosenberger) [inline job-application-tracker-api/src/users/users.service.ts]: Change HttpStatus to HttpStatus.NOT_FOUND (404)

File: job-application-tracker-api/src/users/users.service.ts [hunk 6]
```diff
@@ -68,4 +88,17 @@ export class UsersService {
...
+        }
+        
+        const isPasswordCorrect= await compare(loginDto.password,user.passwordHash);
+        if(!isPasswordCorrect){
+            throw new HttpException('Incorrect password',HttpStatus.UNPROCESSABLE_ENTITY);
```
Comment (ido-rosenberger) [inline job-application-tracker-api/src/users/users.service.ts]: Change HttpStatus to HttpStatus.BAD_REQUEST (400)

File: job-application-tracker-api/src/users/dto/user.dto.ts [hunk 7]
```diff
@@ -0,0 +1,13 @@
+export class User{
```
Comment (ido-rosenberger) [inline job-application-tracker-api/src/users/dto/user.dto.ts]: File should be under a "Models" folder as this is not a DTO but a domain model (also rename file name)

File: job-application-tracker-api/src/users/users.controller.ts [hunk 8]
```diff
@@ -4,6 +4,8 @@ import { Body, Controller, Delete, Get, Param, Patch, Post,
 import { UsersService } from './users.service';
```
Comment (ido-rosenberger) [inline job-application-tracker-api/src/users/users.controller.ts:7]: Reformat this file tfi wtf

File: job-application-tracker-api/src/users/users.service.ts [hunk 9]
```diff
@@ -68,4 +88,17 @@ export class UsersService {
...
+        const user = await this.users.find(user => user.email===loginDto.email);
+        if(!user){
+            throw new HttpException('User not found',HttpStatus.UNPROCESSABLE_ENTITY);
+        }
+        
+        const isPasswordCorrect= await compare(loginDto.password,user.passwordHash);
+        if(!isPasswordCorrect){
+            throw new HttpException('Incorrect password',HttpStatus.UNPROCESSABLE_ENTITY); 
+        }
+        return user;
```
Comment (ido-rosenberger) [inline job-application-tracker-api/src/users/users.service.ts]: Dont return the user, you need to create a jwt token and return it

**Review by ido-rosenberger — CHANGES_REQUESTED** [review]

File: job-application-tracker-api/src/users/dto/user.dto.ts (hunk 7 above)
Comment (tomervak) [reply job-application-tracker-api/src/users/dto/user.dto.ts]: this is  זמני!

File: job-application-tracker-api/src/users/users.service.ts [hunk 10]
```diff
@@ -1,65 +1,101 @@
...
             "email": "bonfil98@gmail.com",
-            "password": "123456"
+            "passwordHash": "$2a$10$oY3pF8WCLCsBMYFpSeTM.uld80NhHAnX797aasFRZbL1s8hxDMbxS",
+            "roles":[Role.admin,Role.user]
         }
-    ]
+    ];
 
+    //TODO: replace with a mapper
+    getUserResponse(user: User) :UserDto {
```
Comment (ido-rosenberger) [inline job-application-tracker-api/src/users/users.service.ts]: Create a file under users folder named ```user-mapper.ts```. Write a class called ```UserMapper``` and add a static function called ```toUserDto(user: User)``` and move this logic there.

**Review by ido-rosenberger — APPROVED** [review]

**Pull request merged**